### ISSUE TITLE
Apply is not included in oc

### ISSUE THREAD

**Issue by smarterclayton** [openshift/origin#6138]
It needs to be

**Comment by liggitt** [comment]
There are several upstream fixes needed before it can be enabled

**Comment by smarterclayton** [comment]
[Citation needed]

**Comment by deads2k** [comment]
I'm pretty sure that the resolution to https://github.com/kubernetes/kubernetes/issues/16543 about `apply` was 

> That said, if we can't find a clean and simple way to use their historical context when the version changes, then maybe it is the best experience we can offer. At this point, I'm tempted to do just that, since this feature is dragging out a long time.

If that's as good as it gets, then feature is kind of borked across API versions and is heavily client-logic dependent, and may decide to simply stop performing the diff it was supposed to provide.

**Comment by smarterclayton** [comment]
Being broken identically to Kube is ok.

On Dec 1, 2015, at 9:12 AM, David Eads <notifications@github.com> wrote:

I'm pretty sure that the resolution to kubernetes/kubernetes#16543
<https://github.com/kubernetes/kubernetes/issues/16543> about apply was

That said, if we can't find a clean and simple way to use their historical
context when the version changes, then maybe it is the best experience we
can offer. At this point, I'm tempted to do just that, since this feature
is dragging out a long time.

If that's as good as it gets, then feature is kind of borked across API
versions and is heavily client-logic dependent, and may decide to simply
stop performing the diff it was supposed to provide.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/6138#issuecomment-160980078>.

**Comment by deads2k** [comment]
> Being broken identically to Kube is ok.

It's sort of like exposing the "this resource will self-destruct upon the next API version" button in our API.  Do we really want that?

@jwforres if we do this and then allow mutation in the web console, you'll need to keep track of exactly how this magic annotation is set based on "user intent" and be very careful never to change the API version used to encode the annotation or you'll accidentally cause all previous state used by the `apply` operation to fail.

The logic is here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/apply.go#L83, but its essentially a diff between levels that attempts to build and resolve conflicts between strategic merge patches for user changes to a particular resource.

**Comment by liggitt** [comment]
the web console doesn't touch annotations on existing objects

**Comment by deads2k** [comment]
@jwforres oh, and you'll need to sort out how to either handle skewed object conversions client-side or at least detect it sufficiently to know that you need to refetch from the server in a particular format.  Keep in mind that will mean that you'll to be able to edit a particular version of the resource to be safe from conversion defaulting.

**Comment by deads2k** [comment]
> the web console doesn't touch annotations on existing objects

In which case, the combination of `apply` and mutating web console is hosed.

**Comment by liggitt** [comment]
> In which case, the combination of apply and mutating web console is hosed.

yes

**Comment by deads2k** [comment]
> In which case, the combination of apply and mutating web console is hosed.
> yes

Not just in the "this is hard to use" sense, but in the "if you accidentally touch one of these things, all your changes will be erased and you make actually lose the entire resource" sense.

It's a self-destruct button.

**Comment by smarterclayton** [comment]
We make a general statement that we extend kubectl.  We can acknowledge
certain features are dangerous or have flaws from upstream Kube as long as
their use is valuable to end users.  Saying we don't include something from
Kube because we don't agree with how it is implemented is a much stronger
statement that I would like to reserve for very special situations (which
I'm not sure this qualifies).  The correct place to fix apply is upstream.

If a user wants to use apply, they should be able to use apply.  Using
apply may mean they need to acknowledge that certain things may not work.
I may not have been following all the issues, it sounds like there are two:

1. API versions changing and old conversions not being available breaks
apply - theoretical, we can solve that problem in three years when we stop
supporting V1.
2. Edits in the UI are lost?? - I don't understand this, apply should be
doing a three way merge.

Others?


On Dec 1, 2015, at 9:27 AM, David Eads <notifications@github.com> wrote:

In which case, the combination of apply and mutating web console is hosed.
yes

Not just in the "this is hard to use" sense, but in the "if you
accidentally touch one of these things, all your changes will be erased and
you make actually lose the entire resource" sense.

It's a self-destruct button.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/6138#issuecomment-160983387>.

**Comment by deads2k** [comment]
> 2. Edits in the UI are lost?? - I don't understand this, apply should be
> doing a three way merge.

Yeah, you'd think it would.  However, `apply` requires every client that ever makes a change to know exactly how to patch the new changes into the `apply` annotation.  If that annotation isn't built properly, the three way merge in `apply` doesn't work and changes can be stomped incorrectly.

`apply` requires every client that exists to know exactly how a particular client will use unstructured data or it doesn't work.

**Comment by smarterclayton** [comment]
So apply has a bug, we warn users, and work upstream on that.


On Dec 1, 2015, at 9:43 AM, David Eads <notifications@github.com> wrote:


   1. Edits in the UI are lost?? - I don't understand this, apply should be
   doing a three way merge.

Yeah, you'd think it would. However, apply requires every client that ever
makes a change to know exactly how to patch the new changes into the apply
annotation. If that annotation isn't built properly, the three way merge in
apply doesn't work and changes can be stomped incorrectly.

apply requires every client that exists to know exactly how a particular
client will use unstructured data or it doesn't work.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/6138#issuecomment-160986878>.
